cmd/utils/flags.go

```diff
@@ -2487,6 +2486,49 @@ func MakeConsolePreloads(ctx *cli.Context) []string {
 	return preloads
 }
 
+// ResolveStateScheme resolve state scheme from CLI flag, config file and persistent state.
+//
+// 1. If config isn't provided, write hash mode to config by default, so in current function, config is nonempty.
+// 2. If persistent state and cli is empty, use config param.
+// 3. If persistent state is empty, provide CLI flag and config, choose CLI to return.
+// 4. If persistent state is nonempty and CLI isn't provided, persistent state should be equal to config.
+// 5. If all three items are provided: if any two of the three are not equal, return error.
+func ResolveStateScheme(ctx *cli.Context, stateSchemeCfg string, disk ethdb.Database) (string, error) {
+	stored := rawdb.ReadStateScheme(disk)
+	if stored == "" {
+		// there is no persistent state data in disk db(e.g. geth init)
+		if !ctx.IsSet(StateSchemeFlag.Name) {
+			log.Info("State scheme set by config", "scheme", stateSchemeCfg)
+			return stateSchemeCfg, nil
+		}
+		// if both CLI flag and config are set, choose CLI
+		scheme := ctx.String(StateSchemeFlag.Name)
+		if !ValidateStateScheme(scheme) {
+			return "", fmt.Errorf("invalid state scheme param in CLI: %s", scheme)
+		}
+		log.Info("State scheme set by CLI", "scheme", scheme)
+		return scheme, nil
+	}
+	if !ctx.IsSet(StateSchemeFlag.Name) {
+		if stored != stateSchemeCfg {
+			return "", fmt.Errorf("incompatible state scheme, stored: %s, config: %s", stored, stateSchemeCfg)
```

Comment:
should not return error if stateSchemeCfg is empty

Reply:
ditto. `stateSchemeCfg` cannot be empty in `ResolveStateScheme` function.